File: leet_BS_0033_search_rotated_sorted_array.cpp
class Solution {
public:
    int search(vector<int>& nums, int target)
    {
        int n = nums.size();
        int R = n - 1;
        int L = 0;
        if (target == nums[L])
            return L ;
        if (target == nums[R])
            return R ;
        while (L < R)
        {
            int mid = (R - L) / 2 + L;
            if (target == nums[mid])
                return mid ;
            if (nums[0] < nums[mid])
            {
                if (nums[0] < target && target < nums[mid])
                    R = mid;
                else
                    L = mid + 1;
            }
            else
            {
                if (nums[mid] < target && target < nums[n - 1])
                    L = mid + 1;
                else
                    R = mid;
            }
        }
        return -1 ;
    }
};

File: leet_BS_0033_search_rotated_sorted_array.ts
function search(nums: number[], target: number): number {
    let len = nums.length
    let r = len - 1
    let l = 0
    while (l <= r) {
        let mid = Math.floor((r - l) / 2) + l
        if (nums[mid] == target) {
            return mid
        }
        if (nums[l] < nums[mid]) {
            if (nums[l] <= target && nums[mid] >= target) {
                r = mid - 1
            } else {
                l = mid + 1
            }
        } else {
            if (nums[r] >= target && nums[mid] <= target) {
                l = mid + 1
            } else {
                r = mid - 1
            }
        }
    }
    return -1
};
